File: v2/src/main/mongo/schema.ts
import type { MongoClient } from 'mongodb';
import type { SchemaField, SchemaReport } from '../../shared/schema.js';

const MAX_DISTINCT_TRACKED = 64;

interface Stat {
  presenceCount: number;
  types: Map<string, number>;
  values: Set<string>;
}

export async function analyzeSchema(
  client: MongoClient,
  db: string,
  collection: string,
  sampleSize = 1000,
): Promise<SchemaReport> {
  const coll = client.db(db).collection(collection);
  const totalDocs = await coll.estimatedDocumentCount();
  const docs = await coll
    .aggregate([{ $sample: { size: Math.min(sampleSize, Math.max(totalDocs, 1)) } }])
    .toArray();

  const stats = new Map<string, Stat>();
  for (const doc of docs) walk(doc as Record<string, unknown>, '', stats);

  const fields: SchemaField[] = [...stats.entries()]
    .map(([path, stat]) => {
      const totalTypeHits = [...stat.types.values()].reduce((s, n) => s + n, 0) || 1;
      const types: Record<string, number> = {};
      for (const [t, n] of stat.types) types[t] = n / totalTypeHits;
      return {
        path,
        presence: docs.length === 0 ? 0 : stat.presenceCount / docs.length,
        types,
        approxDistinct: stat.values.size,
      };
    })
    .sort((a, b) => b.presence - a.presence);

  return { sampleSize: docs.length, totalDocs, fields };
}

function walk(value: unknown, prefix: string, stats: Map<string, Stat>): void {
  if (value === undefined) return;
  if (value === null) {
    bump(stats, prefix, 'null', value);
    return;
  }
  if (Array.isArray(value)) {
    bump(stats, prefix, 'array', value);
    if (value.length > 0) walk(value[0], `${prefix}[]`, stats);
    return;
  }
  if (typeof value === 'object') {
    if (isBsonWrapper(value)) {
      bump(stats, prefix, bsonWrapperType(value as Record<string, unknown>), value);
      return;
    }
    bump(stats, prefix, 'object', value);
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      walk(v, prefix ? `${prefix}.${k}` : k, stats);
    }
    return;
  }
  if (typeof value === 'string') bump(stats, prefix, 'string', value);
  else if (typeof value === 'number') bump(stats, prefix, 'number', value);
  else if (typeof value === 'boolean') bump(stats, prefix, 'boolean', value);
}

function bump(
  stats: Map<string, Stat>,
  path: string,
  type: string,
  value: unknown,
): void {
  if (!path) return;
  let stat = stats.get(path);
  if (!stat) {
    stat = { presenceCount: 0, types: new Map(), values: new Set() };
    stats.set(path, stat);
  }
  stat.presenceCount++;
  stat.types.set(type, (stat.types.get(type) ?? 0) + 1);
  if (stat.values.size < MAX_DISTINCT_TRACKED) {
    stat.values.add(typeof value === 'object' ? JSON.stringify(value) : String(value));
  }
}

function isBsonWrapper(v: unknown): boolean {
  if (typeof v !== 'object' || v === null) return false;
  const cls = v.constructor?.name ?? '';
  return ['ObjectId', 'Long', 'Decimal128', 'Binary', 'Timestamp', 'BSONRegExp', 'MinKey', 'MaxKey', 'Date'].includes(cls);
}

function bsonWrapperType(v: Record<string, unknown>): string {
  const cls = (v as { constructor: { name: string } }).constructor.name;
  switch (cls) {
    case 'ObjectId': return 'objectid';
    case 'Long': return 'long';
    case 'Decimal128': return 'decimal';
    case 'Binary': return 'binary';
    case 'Timestamp': return 'timestamp';
    case 'BSONRegExp': return 'regex';
    case 'MinKey': return 'minkey';
    case 'MaxKey': return 'maxkey';
    case 'Date': return 'date';
    default: return cls.toLowerCase();
  }
}

File: v3/src/main/kotlin/io/mex/mongo/Schema.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document
import org.bson.types.Binary
import org.bson.types.Decimal128
import org.bson.types.MaxKey
import org.bson.types.MinKey
import org.bson.types.ObjectId
import java.util.Date

private const val MAX_DISTINCT = 64

data class SchemaField(
    val path: String,
    val presence: Double,        // 0-1
    val types: Map<String, Double>, // sums to 1
    val approxDistinct: Int,
    /** Sampled distinct values, retained only for low-cardinality fields (autocomplete). */
    val sampleValues: List<String> = emptyList(),
) {
    /** The type that dominates this field across the sample, for display next to a suggestion. */
    val dominantType: String get() = types.maxByOrNull { it.value }?.key ?: "mixed"
}

data class SchemaReport(val sampleSize: Int, val totalDocs: Long, val fields: List<SchemaField>)

private class Stat {
    var presenceCount = 0
    val types = HashMap<String, Int>()
    val values = HashSet<String>()
}

fun analyzeSchema(client: MongoClient, db: String, coll: String, sampleSize: Int = 1000): SchemaReport {
    val collection = client.getDatabase(db).getCollection(coll)
    val total = collection.estimatedDocumentCount()
    val sz = minOf(sampleSize, maxOf(total.toInt(), 1))
    val docs = collection.aggregate(listOf(Document("\$sample", Document("size", sz)))).toList()
    val stats = HashMap<String, Stat>()
    for (doc in docs) walk(doc, "", stats)
    val fields = stats.entries.map { (path, stat) ->
        val total = stat.types.values.sum().coerceAtLeast(1)
        SchemaField(
            path = path,
            presence = if (docs.isEmpty()) 0.0 else stat.presenceCount.toDouble() / docs.size,
            types = stat.types.mapValues { it.value.toDouble() / total },
            approxDistinct = stat.values.size,
            // Enumerable fields get their values kept so the query editor can suggest them;
            // anything near the distinct cap is effectively free-form and not worth listing.
            sampleValues = if (stat.values.size < MAX_DISTINCT) stat.values.sorted() else emptyList(),
        )
    }.sortedByDescending { it.presence }
    return SchemaReport(sampleSize = docs.size, totalDocs = total, fields = fields)
}

private fun walk(value: Any?, prefix: String, stats: MutableMap<String, Stat>) {
    if (value == null) {
        bump(stats, prefix, "null", null); return
    }
    when (value) {
        is List<*> -> {
            bump(stats, prefix, "array", value)
            value.firstOrNull()?.let { walk(it, "$prefix[]", stats) }
        }
        is Document -> {
            bump(stats, prefix, "object", value)
            for ((k, v) in value) walk(v, if (prefix.isEmpty()) k else "$prefix.$k", stats)
        }
        is String -> bump(stats, prefix, "string", value)
        is Boolean -> bump(stats, prefix, "boolean", value)
        is Number -> bump(stats, prefix, "number", value)
        is ObjectId -> bump(stats, prefix, "objectid", value)
        is Date -> bump(stats, prefix, "date", value)
        is Decimal128 -> bump(stats, prefix, "decimal", value)
        is Binary -> bump(stats, prefix, "binary", value)
        is MinKey -> bump(stats, prefix, "minkey", value)
        is MaxKey -> bump(stats, prefix, "maxkey", value)
        else -> bump(stats, prefix, value::class.java.simpleName.lowercase(), value)
    }
}

private fun bump(stats: MutableMap<String, Stat>, path: String, type: String, v: Any?) {
    if (path.isEmpty()) return
    val s = stats.getOrPut(path) { Stat() }
    s.presenceCount++
    s.types[type] = (s.types[type] ?: 0) + 1
    if (s.values.size < MAX_DISTINCT) s.values.add(v?.toString() ?: "null")
}
